lib/src/rules/no_constant_pattern_type_literal.dart

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r"Don't use constant patterns with type literals.";
+
+const _details = r'''
+Use `== TypeName` or `TypeName _` instead of type literals in patterns.
+
+**BAD:**
+```dart
+void f(Type x) {
+  if (x case int) {
+    print('int');
+  }
+}
+```
+
+**GOOD:**
+```dart
+void f(Type x) {
+  if (x case == int) {
+    print('int');
+  }
+}
+```
+
+**BAD:**
+```dart
+void f(Object? x) {
+  if (x case int) {
+    print('int');
+  }
+}
+```
+
+**GOOD:**
+```dart
+void f(Object? x) {
+  if (x case int _) {
+    print('int');
+  }
+}
+```
+
+''';
+
+class NoConstantPatternTypeLiteral extends LintRule {
```

Comment:
I think I'd prefer `NoConstantPatternTypeLiterals` --- on balance we have more plural thank singular names.

Reply:
With the comment from Brian below, I renamed it to `TypeLiteralInConstantPattern`, to be consistent with the `type_literal_in_constant_pattern` name. Is this a correct way to name lint rules?

Reply:
Yes. It's easier to find the implementation if the class name and lint name match.